### PR TITLE
Implement Lift using interners instead of in_arena

### PR DESCRIPTION
**Pull request by Zoxc**
r? @eddyb
cc @cjgillot

### CODE REVIEW

File: src/librustc_data_structures/sharded.rs [hunk 1]
```diff
@@ -137,6 +137,20 @@ impl<K: Eq + Hash + Copy> ShardedHashMap<K, ()> {
     }
 }
 
+pub unsafe trait IntoPointer {
```
Comment (Centril) [inline src/librustc_data_structures/sharded.rs]: Documenting the proof obligations would be good.

Comment (Zoxc) [reply src/librustc_data_structures/sharded.rs]: I'll just make it safe instead =P

File: src/librustc/arena.rs [hunk 2]
```diff
@@ -123,6 +123,9 @@ macro_rules! arena_types {
             [few] inferred_outlives_crate: rustc::ty::CratePredicatesMap<'tcx>,
             [] upvars: rustc_data_structures::fx::FxIndexMap<rustc_hir::HirId, rustc_hir::Upvar>,
 
+            // Interned types
+            [] tys: rustc::ty::TyS<$tcx>,
```
Comment (eddyb) [inline src/librustc/arena.rs]: Huh, so we're not using the `DroplessArena` for these anymore?

File: src/librustc_typeck/astconv.rs [hunk 3]
```diff
@@ -1890,10 +1890,10 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
     fn one_bound_for_assoc_type<I>(
         &self,
         all_candidates: impl Fn() -> I,
-        ty_param_name: &str,
+        ty_param_name: impl Fn() -> String,
         assoc_name: ast::Ident,
         span: Span,
-        is_equality: Option<String>,
+        is_equality: impl Fn() -> Option<String>,
```
Comment (eddyb) [inline src/librustc_typeck/astconv.rs]: Am I missing something, or are the changes in this file unrelated?

**Comment by eddyb** [comment]
r=me with https://github.com/rust-lang/rust/pull/67791#discussion_r366912473 explained or resolved

File: src/librustc/arena.rs (hunk 2 above)
Comment (Zoxc) [reply src/librustc/arena.rs]: We do, but `TyS` is not `Copy`, so we have to declare it.

**Pull request merged**